### PR TITLE
IDEMPIERE-5723 NPE when allocate has both AR and AP invoice

### PR DESCRIPTION
**Pull request by hieplq**
base on patch from gauravsontakke

# Pull Request Checklist

- [x] My code follows the [code guidelines](https://wiki.idempiere.org/en/Contributing_to_Trunk) of this project
- [ ] My code follows the best practices of this project
- [x] I have performed a self-review of my own code
- [x] My code is easy to understand and review. 
- [x] I have checked my code and corrected any misspellings
- [ ] In hard-to-understand areas, I have commented my code.
- [x] My changes generate no new warnings
### Tests
- [x] I have tested the direct scenario that my code is solving
- [ ] I checked all collaterals that can be affected by my changes, and tested other potential affected scenarios
- [x] New and existing unit tests pass locally with my changes
- [x] I have added unit tests that prove my fix is effective or that my feature works
### Documentation
- [ ] I have made corresponding changes to the documentation as follows:
- - [ ] New feature (non-breaking change which adds functionality): I have created the New Feature page in the project wiki explaining the functionality and how to use it. If relevant, I have committed sample data to the core seed to have usable examples in GardenWorld.
- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected): I have documented the change in a clear way that everyone in the community can understand the impact of the change.
- - [ ] Improvement (improves and existing functionality): This documentation is needed if the improvement changes the way the user interacts with the system or the outcome of a process/task changes. If it is just, for instance, a performance improvement, documentation might not be needed. 
- [ ] The changed/added documentation is in the project wiki (not privately-hosted pdf files or links pointing to a company website) and is complete and self-explanatory.

### CODE REVIEW

File: org.idempiere.test/src/org/idempiere/test/model/AllocationTest.java [hunk 1]
```diff
@@ -170,6 +170,67 @@ public void testAllocateCharge() {
...
+
+			completeDocument(alloc);
+			postDocument(alloc);
+			
+			alloc.load(getTrxName());
+			
+			assertTrue(alloc.isPosted(), "Allocation not posted");
+			
+		}finally {
+			rollback();
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/model/AllocationTest.java:229]: this is harmless but not necessary. trx always rollback for each test.

File: org.adempiere.base/src/org/compiere/acct/Doc_AllocationHdr.java [hunk 2]
```diff
@@ -1065,26 +1072,30 @@ allocationSource, getC_Currency_ID(),
...
-			DocLine_Allocation line = (DocLine_Allocation)p_lines[i];			
-			if (line.getC_Invoice_ID() != 0)
-			{
+			DocLine_Allocation line = (DocLine_Allocation)p_lines[i];
+			
+			if (line.getC_Invoice_ID() == 0)
+				continue;
+
+			if (invList.containsKey(line.getC_Invoice_ID())){
+				log.severe(line.getC_Invoice_ID() + ":same invoice on multi allocation line");
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/acct/Doc_AllocationHdr.java:1088]: "same invoice included in more than one allocation line" sounds clearer

File: org.adempiere.base/src/org/compiere/acct/Doc_AllocationHdr.java [hunk 3]
```diff
@@ -1065,26 +1072,30 @@ allocationSource, getC_Currency_ID(),
 	 *	@param acct account
 	 *	@return Error Message or null if OK
 	 */
-	private String createInvoiceRoundingCorrection (MAcctSchema as, Fact fact, MAccount acct) 
+	private String createInvoiceRoundingCorrection (MAcctSchema as, Fact fact, MAccount acctAr, MAccount acctAp) 
 	{
-		ArrayList<MInvoice> invList = new ArrayList<MInvoice>();
+		Hashtable<Integer, MInvoice> invList = new Hashtable<>();
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/acct/Doc_AllocationHdr.java:1077]: Map and HashMap should be preferred over hashtable here.

File: org.idempiere.test/src/org/idempiere/test/model/AllocationTest.java [hunk 4]
```diff
@@ -170,6 +170,67 @@ public void testAllocateCharge() {
...
+	/**
+	 * https://idempiere.atlassian.net/browse/IDEMPIERE-5723
+	 */
+	public void testAllocateInvoiceArAp() {
+		MBPartner bpartner = MBPartner.get(Env.getCtx(), DictionaryIDs.C_BPartner.JOE_BLOCK.id); 
+		Timestamp currentDate = Env.getContextAsDate(Env.getCtx(), "#Date");
+		
+		Calendar cal = Calendar.getInstance();
+		cal.setTimeInMillis(currentDate.getTime());
+		Timestamp date = new Timestamp(cal.getTimeInMillis());
```
Comment (hengsin) [inline org.idempiere.test/src/org/idempiere/test/model/AllocationTest.java:183]: TimeUtil.getDay(null) give you today's date

File: org.adempiere.base/src/org/compiere/acct/Doc_AllocationHdr.java [hunk 5]
```diff
@@ -460,6 +466,7 @@ else if (isUsingClearing && line.getC_CashLine_ID() != 0) // Avoid usage of clea
 				&& (getC_Currency_ID() != as.getC_Currency_ID()			//	payment allocation in foreign currency
 					|| getC_Currency_ID() != line.getInvoiceC_Currency_ID()))	//	allocation <> invoice currency
 			{
+				//do need a check here for as.isAccrual()?
```
Comment (hieplq) [inline org.adempiere.base/src/org/compiere/acct/Doc_AllocationHdr.java:469]: @hengsin @CarlosRuiz-globalqss 
wdyt about this check? just uncheck "Accrual" on "Accounting Schema" and run unit test for allocation some not pass by NPE at createInvoiceGainLoss

image

Comment (CarlosRuiz-globalqss) [reply org.adempiere.base/src/org/compiere/acct/Doc_AllocationHdr.java:470]: Well, seems like a different bug, not to mix with this ticket.

Cash based accounting (non-accrual) is rarely used, that must be the reason nobody has caught such error.

Comment (CarlosRuiz-globalqss) [reply org.adempiere.base/src/org/compiere/acct/Doc_AllocationHdr.java:470]: Ah @hieplq - also answering your question (I didn't notice at first) - most tests are designed to work in a vanilla seed GardenWorld - some tests will fail if the GW configuration is modified, or even if there is additional data.

Not sure if that's something we must solve generically in all tests, but sounds like big work.

Comment (hengsin) [reply org.adempiere.base/src/org/compiere/acct/Doc_AllocationHdr.java:470]: hi @hieplq , that test is for accrual accounting only. I guess the fix is pretty simple for this one, we just have to either make sure accounting schema is accrual accounting or skip the test if accounting schema is non-accrual.

**Review by hengsin — APPROVED** [review]
pass unit test

**Pull request merged**